File: Algorithms/Data Structure dependent/Trees/flatten.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, TreeNode* ans){
        if (root == nullptr) return;
        TreeNode cur(root->val);
        ans->right = &cur;
        ans = ans->right;
        cout << ans->val << ",";
        helper(root->left, ans);
        helper(root->right, ans);
    }
    
    void flatten(TreeNode* root) {
        TreeNode start(-1);
        helper(root, &start);
        auto next = start.right;
        root->left = nullptr;
        root->right = next->right;
    }
};

File: Algorithms/Data Structure dependent/LinkedList/flatten.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Prev *Node
 *     Next *Node
 *     Child *Node
 * }
 */
func helper(root *Node) *Node {
    if root == nil || (root.Child == nil && root.Next == nil){
        return root
    }
    if root.Child != nil {
        tmp := root.Next
        root.Next = root.Child
        /* Get last node from flattened list and set their next to our old next*/
        lastChild := helper(root.Child)
        root.Child = nil
        lastChild.Next = tmp
        helper(lastChild)
    }

    return helper(root.Next)
}
func flatten(root *Node) *Node {
    helper(root)
    prev := (*Node)(nil)
    cur := root
    for cur != (*Node)(nil) {
        cur.Prev = prev
        prev = cur
        cur = cur.Next
    }
    return root
}